### PR TITLE
task(fxa-settings): rework Two-step auth unit row

### PR DESCRIPTION
**Pull request by meandavejustice**
## Because

- detect Two-step auth and add ability to disable

## This pull request

- created `UnitRowTwoStepAuth` component
- added `SecondaryButtonClassName`,` SecondaryCta`, and `SecondaryModalReveal`
options to UnitRow
- updated Security component, and all tests and stories to reflect
changes
- updated each story touched with `<LocationProvider/>` fix

## Issue that this pull request solves

Closes: #4913 #6327 

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

unset-totp

set-totp

### CODE REVIEW

File: packages/fxa-settings/src/components/UnitRow/index.tsx
Comment (LZoog) [inline packages/fxa-settings/src/components/UnitRow/index.tsx:38]: This looks good in Storybook and all the way up to tablet:
image


But at desktop, they do this: 
image

I think we can fix this by wrapping each button in a `div` so they take up equal space in `unit-row-actions` instead of wrapping both buttons in a `div`.

Also, I think we need a new class that's not `cta-neutral`, that one's for grey buttons and the hover state is a little off. We need a new one for "caution" (that's what Figma calls it anyway), you could add them to `ctas.scss`:
image

image

File: packages/fxa-settings/src/components/UnitRow/index.stories.tsx
Comment (LZoog) [inline packages/fxa-settings/src/components/UnitRow/index.stories.tsx:65]: I think we should probably call these `[secondaryModalRevealed, revealSecondaryModal, hideSecondaryModal]` just so it's clear these are all for the modal tied to the secondary CTA.

File: packages/fxa-settings/src/components/UnitRow/index.stories.tsx
Comment (LZoog) [inline packages/fxa-settings/src/components/UnitRow/index.stories.tsx:71]: It looks like `secondaryCtaText` is set to "Disable" by default, so I think you can take this out.

File: packages/fxa-settings/src/components/UnitRow/index.test.tsx
Comment (LZoog) [inline packages/fxa-settings/src/components/UnitRow/index.test.tsx:94]: I think you can take `secondaryCtaText="Disable"` out since it gets "Disable" by default.

I also think `ctaText` can be removed, it should get "Create" when `headerValue` is falsey.

File: packages/fxa-settings/src/components/UnitRow/index.tsx
Comment (LZoog) [inline packages/fxa-settings/src/components/UnitRow/index.tsx:63]: Looks like `alertBarRevealed` is no longer used here.

File: packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx
Comment (LZoog) [inline packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx:57]: I think we can simplify this a tad since `noHeaderValueText` is also a prop here and there's a lot of conditionals based on `exists`. What do you think about something like:

```
  const conditionalUnitRowProps = exists ? {
    headerValueClassName: 'text-green-800',
    headerValue: 'Enabled',
    secondaryButtonClassName: 'bg-red-500 text-white border-red-600',
    revealSecondaryModal: revealModal,
  } : {
    headerValue: null,
    noHeaderValueText: 'Not Set',
    secondaryCtaText: undefined,
    revealSecondaryModal: undefined
  };

  return (
    <UnitRow
      header="Two-step authentication"
      route="/beta/settings/two_step_authentication"
      {...conditionalUnitRowProps}
    >
```

File: packages/fxa-settings/src/components/UnitRow/index.tsx
Comment (LZoog) [inline packages/fxa-settings/src/components/UnitRow/index.tsx:12]: We probably want to call this `className` just for consistency.

File: packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx
Comment (dannycoates) [inline packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx:79]: "two-step" ... also lower case "authentication"

File: packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx
Comment (dannycoates) [inline packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx:46]: should be `{...existing, totp: { exists: false, verified: false }}`

File: packages/fxa-settings/src/components/UnitRow/index.tsx
Comment (jodyheavener) [reply packages/fxa-settings/src/components/UnitRow/index.tsx:38]: Adding to this, the mobile layout needs some margin work.

Screen Shot 2020-08-26 at 2 21 31 PM

File: packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx
Comment (jodyheavener) [inline packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx:79]: The spelling/capitalization needs to be updated, but also this modal should more closely resemble the Figma doc. Looks like we need to use the Red disable button that says "Disable", instead of a blue button that says "Confirm", and the "replacing your recovery codes" link looks off. The button horizontal spacing is also off, but I am noticing this on main too, so it could be addressed later if not here.

Figma:

Screen Shot 2020-08-26 at 2 27 13 PM

Actual:

Screen Shot 2020-08-26 at 2 26 43 PM

File: packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx
Comment (meandavejustice) [reply packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx:75]: lol, copy and paste

File: packages/fxa-settings/src/components/UnitRow/index.tsx
Comment (meandavejustice) [reply packages/fxa-settings/src/components/UnitRow/index.tsx:63]: It needs to stay, unfortunately since it is passed through to `ModalButton`

Comment (jodyheavener) [reply packages/fxa-settings/src/components/UnitRow/index.tsx:63]: Maybe it should be, but currently it's not being used 🤔 
Screen Shot 2020-08-27 at 1 34 01 PM

File: packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx
Comment (meandavejustice) [reply packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx:79]: I updated the text of and color of the disable button, I think we should file a new bug for fixing the default styles of the modal since it affects other parts of the app, and seems out of scope for this issue. Filed https://jira.mozilla.com/browse/FXA-2486

File: packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx
Comment (jodyheavener) [inline packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx:98]: This class makes the link awkward. You can remove `text-sm mt-2`. Additionally I think you'll need to add `{' '}` right before the opening LinkExternal tag to force a space.
Screen Shot 2020-08-27 at 2 40 22 PM

File: packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx
Comment (jodyheavener) [inline packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx:91]: Lowercase `two`

File: packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx
Comment (jodyheavener) [inline packages/fxa-settings/src/components/UnitRowTwoStepAuth/index.tsx:110]: Oh! `AlertBar` now supports a `type` -- this could be `success`, `error`, or `info`. Could we update this to use the type for when the update was successful or failed?

File: packages/fxa-settings/src/components/Modal/index.tsx
Comment (jodyheavener) [inline packages/fxa-settings/src/components/Modal/index.tsx:87]: I just noticed this uses `confirmBtnClassName` _and_ `cta-primary`, which would create competing styles. Can `confirmBtnClassName` just default to `cta-primary`?

Comment (meandavejustice) [reply packages/fxa-settings/src/components/Modal/index.tsx:87]: I like that much better :)

**Comment by jodyheavener** [comment]
@meandavejustice now that #6350 is merged you should be able to using the layout I set up for modal buttons. Additionally, since this your PR adds the cta-caution class and the new props to set modal class/text, could you uncomment this? https://github.com/mozilla/fxa/blob/main/packages/fxa-settings/src/components/UnitRowRecoveryKey/index.tsx#L83-L84

File: packages/fxa-settings/src/components/Modal/index.tsx
Comment (jodyheavener) [inline packages/fxa-settings/src/components/Modal/index.tsx:34]: It looks like all modals are the same width in the Figma doc, is there are specific reason for needing varying modal max-widths? It might be best to make them consistent. Looks like `max-w-md` is closest, but if we do that we should reduce the horizontal padding so "Disable two-step authentication?" doesn't break to two lines

File: packages/fxa-settings/src/components/UnitRow/index.tsx
Comment (jodyheavener) [inline packages/fxa-settings/src/components/UnitRow/index.tsx:37]: I think this is missing a `cta-neutral` class:

Screen Shot 2020-09-08 at 3 06 07 PM

**Review by jodyheavener — COMMENTED** [review]
🤠 Yeehaw this is looking and feeling great. Mutation works on my end. Just a couple cosmetic comments and then I think we should be good to ship.

File: packages/fxa-settings/src/components/UnitRow/index.tsx
Comment (meandavejustice) [reply packages/fxa-settings/src/components/UnitRow/index.tsx:97]: This is what we decided because of the button breakage on mobile and certain wide desktop widths, I'm happy to move it back so long as it doesn't break again

Comment (meandavejustice) [reply packages/fxa-settings/src/components/UnitRow/index.tsx:97]: It appears to be working well now, had to remove some `ml-2` classes on the buttons we added when trying to get it working before.

Comment (jodyheavener) [reply packages/fxa-settings/src/components/UnitRow/index.tsx]: I'm not fully up to date on the specific issue, but if it persists with that class feel free to ping me with a description of what's going on and I'll see what I can do -- I love that kind of stuff

Comment (jodyheavener) [reply packages/fxa-settings/src/components/UnitRow/index.tsx]: It looks like this change resulted in a wonky layout for desktop:

Screen Shot 2020-09-09 at 1 35 03 PM

And while there's nothing visually wrong with this mobile layout, it doesn't match the Figma doc:

Screen Shot 2020-09-09 at 1 34 57 PM

However, this looks like a relatively easy fix. Check out this patch, it should do the trick:

```diff
diff --git a/packages/fxa-settings/src/components/UnitRow/index.tsx b/packages/fxa-settings/src/components/UnitRow/index.tsx
index f0687213b..0b43a30ed 100644
--- a/packages/fxa-settings/src/components/UnitRow/index.tsx
+++ b/packages/fxa-settings/src/components/UnitRow/index.tsx
@@ -13,6 +13,7 @@ type ModalButtonProps = {
   revealModal: () => void;
   modalRevealed?: boolean;
   alertBarRevealed?: boolean;
+  leftSpaced?: boolean;
 };
 
 export const ModalButton = ({
@@ -21,6 +22,7 @@ export const ModalButton = ({
   revealModal,
   modalRevealed,
   alertBarRevealed,
+  leftSpaced,
 }: ModalButtonProps) => {
   const modalTriggerElement = useRef<HTMLButtonElement>(null);
   // If the UnitRow children contains an AlertBar that is revealed,
@@ -35,8 +37,9 @@ export const ModalButton = ({
   return (
     <button
       className={classNames(
-        'cta-base cta-neutral transition-standard mobileLandscape:ml-2',
-        className
+        'cta-base transition-standard',
+        leftSpaced && 'ml-2',
+        className || 'cta-neutral'
       )}
       data-testid="unit-row-modal"
       ref={modalTriggerElement}
@@ -81,6 +84,7 @@ export const UnitRow = ({
   ctaText = ctaText || (headerValue ? 'Change' : 'Add');
 
   const location = useLocation();
+  const multiButton = !!(route || secondaryCtaRoute)
 
   return (
     <div className="unit-row">
@@ -98,10 +102,10 @@ export const UnitRow = ({
       </div>
 
       <div className="unit-row-actions">
-        <div>
+        <div className="flex items-center">
           {route && (
             <Link
-              className="cta-neutral cta-base transition-standard mobileLandscape:mr-2"
+              className="cta-neutral cta-base transition-standard mr-2"
               data-testid="unit-row-route"
               to={`${route}${location.search}`}
             >
@@ -110,12 +114,13 @@ export const UnitRow = ({
           )}
 
           {revealModal && (
-            <ModalButton {...{ revealModal, ctaText, alertBarRevealed }} />
+            <ModalButton
+              leftSpaced={multiButton} {...{ revealModal, ctaText, alertBarRevealed }} />
           )}
 
           {secondaryCtaRoute && (
             <Link
-              className="cta-neutral transition-standard mobileLandscape:mr-2"
+              className="cta-neutral cta-base transition-standard mr-2"
               data-testid="unit-row-route"
               to={`${secondaryCtaRoute}${location.search}`}
             >
@@ -125,6 +130,7 @@ export const UnitRow = ({
 
           {revealSecondaryModal && (
             <ModalButton
+              leftSpaced={multiButton}
               revealModal={revealSecondaryModal}
               ctaText={secondaryCtaText}
               className={secondaryButtonClassName}
```

**Comment by jodyheavener** [comment]
@meandavejustice I left you one more followup layout issue comment, but after that I'm ready to hit green 👌

File: packages/fxa-settings/src/components/UnitRow/index.tsx
Comment (meandavejustice) [reply packages/fxa-settings/src/components/UnitRow/index.tsx]: So weird this wasn't showing up for me in storybook, I wonder if I need to be testing on a second monitor, or adjusting my resolution for retina screens. Anyways, I made the changes from the patch and it's looking good in all the states on my machine.

**Review by jodyheavener — APPROVED** [review]
Looking good. Thanks for all the effort you put in to this!

Now that the new `useAlertBar` hook is available ([docs](https://github.com/mozilla/fxa/blob/main/packages/fxa-settings/README.md#alertbar-and-alertexternal)) you have the option to implement it here as well, but I don't want to hold this up any longer than I already have, so it's you're call if you'd like to update before merging, or leave it for later.

**Pull request merged**
